cabot/metricsapp/views/grafana_elastic.py

```diff
@@ -1,88 +1,139 @@
+            context.update(self.get_context_data())
+            return render(self.request, 'metricsapp/grafana_preview_changes.html', context)
 
-        check = ElasticsearchStatusCheck.objects.get(id=kwargs['pk'])
-        check.set_fields_from_grafana(fields)
+        # else preview accepted, continue as usual
+        return super(GrafanaElasticsearchStatusCheckUpdateView, self).form_valid(form)
 
-        return super(GrafanaElasticsearchStatusCheckRefreshView, self).get(self, request, *args, **kwargs)
+    def get_context_data(self, **kwargs):
+        context = super(GrafanaElasticsearchStatusCheckUpdateView, self).get_context_data(**kwargs)
+        if self.object.grafana_panel is not None:
+            context['panel_url'] = self.object.grafana_panel.panel_url
+        return context
+
+
+class GrafanaElasticsearchStatusCheckRefreshView(GrafanaElasticsearchStatusCheckUpdateView):
+    """
+    Variant of CheckUpdateView that pre-fills the form with the latest values from Grafana.
+    Note this requires the session vars by earlier Views in the "create Grafana check" flow to work
+    (GrafanaDashboardSelectView, etc).
+    """
+    def get_form_kwargs(self):
+        """Add kwargs['data'] and fill it with the latest values from Grafana"""
+        kwargs = super(GrafanaElasticsearchStatusCheckRefreshView, self).get_form_kwargs()
+
+        if self.request.method == 'GET':
+            # kwargs['data'] will be the data in the form presented to the user
+            # note we don't use kwargs['initial'] or override get_initial() here because we want to use form.has_changed
+            # later for the preview
+            kwargs['data'] = {}
+
+            # since we are specifying the form data (in order to update values with the latest from Grafana),
+            # we must set all the values or anything we omit will have the default form.widget value
+            # we do this by creating a form and copying the values from it
+            base_form = self.get_form_class()(instance=self.object)
+            for field_name in base_form.fields:
+                kwargs['data'][field_name] = base_form[field_name].value()
+
+            # build fields which will have the latest data from Grafana
+            request = self.request
+            dashboard_info = request.session['dashboard_info']
+            panel_info = request.session['panel_info']
+            series = request.session['series']
+            templating_dict = request.session['templating_dict']
+            grafana_panel = build_grafana_panel_from_session(request.session)
+            grafana_data_source = GrafanaDataSource.objects.get(
+                grafana_source_name=request.session['datasource'],
+                grafana_instance_id=request.session['instance_id']
+            )
+
+            fields = get_status_check_fields(dashboard_info, panel_info, grafana_data_source, templating_dict,
+                                             grafana_panel)
```

Comment:
If we end up adding more data sources with a Grafana integration most of the above would have to be repeated in their views...any way to take it out so we can reuse it?

Reply:
Might be able to set up a base GrafanaRefreshView mixin that only fills get_status_check_fields, then the Elasticsearch-specific one would only do get_es_status_check_fields...will keep thinking about it.